### DISCUSSION TITLE
Why checksums does not have the same behaviors as package and allow 'import' keyword to split

### DISCUSSION THREAD

**Discussion by jmauro** [aquaproj/aqua#2699] [Deprecated (Question and Answer)]
### Question

I was wondering why there is a way to define different path to add package and there is not the mechanism for the checksums? I mean, if you split your packages' configuration, you want to do the same for the checksums, I don't see the point of split the packages files and having a unique big checksums file with all the packages.

### Example Code

```console
$ 
```

Configuration

```yaml

```

**Comment by suzuki-shunsuke** [comment]
Thank you for your feedback.

> I was wondering why there is a way to define different path to add package and there is not the mechanism for the checksums?

This is because we have the motivation to split aqua.yaml but don't have the motivation to split aqua-checksums.json.

About the motivation to split aqua.yaml, we mentioned in the document too.

https://aquaproj.github.io/docs/guides/split-config

> This is useful for CI. You can execute test and lint only when the specific package is updated.

On the other hand, we have no motivation to split aqua-checksums.json.
We don't want to add the unnecessary complexity, so we don't support splitting aqua-checksums.json at the moment.

If you have any motivations to split aqua-checksums.json, please describe the detail.
Why do you want to split aqua-checksums.json?

**Reply by jmauro** [reply]
Hello @suzuki-shunsuke,

Thanks for taking the time to answer.
The checksums of a software should be defined where you define the software you want to use. I don't see why you could import packages from wherever you want and not the checksums associated with it. What's more, `aqua` is even able to update the file where the package is imported from and not the checksum (which awesome by the way)
Let's say I want a master team that provides the default aqua configuration with the following setup:
```
$ cat /.aqua.yml
---
# aqua - Declarative CLI Version Manager
# https://aquaproj.github.io/
checksum:
  enabled: true
  require_checksum: true
  supported_envs:
    - linux
registries:
  - type: standard
    ref: v4.138.0 # renovate: depName=aquaproj/aqua-registry
packages:
  - import: /etc/aqua.d/*/*.yml
  - import: /etc/aqua.d/*/*.yaml
```
So if a new team need new packages they create a new file such as: `/etc/aqua.d/teamA/packages.yaml`, then the checksums file has to be created where the `aqua.yaml` file is (a.k.a `/.aqua-checksums.json`), so when the team updates the package file they have to edit the checksums file, but they might not have edit rights for this file.
This could be annoying. But maybe I understood wrongly the philosophy behind `aqua`.

**Reply by suzuki-shunsuke** [reply]
Thank you for your explanation.

How about splitting aqua.yaml and specifying them by the environment variable `AQUA_GLOBAL_CONFIG`?

e.g.

```
/etc/aqua.d/
  common/
    aqua.yaml
    aqua-checksums.json
  teamA/
    aqua.yaml
    aqua-checksums.json
  teamB/
    aqua.yaml
    aqua-checksums.json
```

The members of team A set `AQUA_GLOBAL_CONFIG` as the following.

```sh
export AQUA_GLOBAL_CONFIG=/etc/aqua.d/teamA/aqua.yaml:/etc/aqua.d/common/aqua.yaml
```

And The members of team B set `AQUA_GLOBAL_CONFIG` as the following.

```sh
export AQUA_GLOBAL_CONFIG=/etc/aqua.d/teamB/aqua.yaml:/etc/aqua.d/common/aqua.yaml
```

**Reply by jmauro** [reply]
In that case, this is means that each team will have to configure their registries as well as theirs global configuration, meaning if teamA disables the checksums' enforcement, the common aqua.yaml will be ignored 😢

**Reply by suzuki-shunsuke** [reply]
So you mean you want to enforce each team to enable the checksum verification, right?
If so, how about this?

```
/etc/aqua.d/
  teamA/
    aqua.yaml # Team A doesn't have the write permission
    imports/ # Team A has the write permission
      tool-a.yaml
      tool-b.yaml
    aqua-checksums.json # Team A has the write permission
```

aqua.yaml

```yaml
checksum:
  enabled: true
  require_checksum: true
  supported_envs:
    - linux
registries:
  - type: standard
    ref: v4.138.0 # renovate: depName=aquaproj/aqua-registry
packages:
  - import: imports/*.yaml
```

In this case, one of issues is that Team A can't update registries themselves because they don't have the write permission of aqua.yaml.

**Reply by suzuki-shunsuke** [reply]
The feature to enforce the checksum verification is interesting and valuable in terms of security.
The feature would be useful for Monorepo.
I created the issue.

- https://github.com/aquaproj/aqua/issues/2702

**Reply by jmauro** [reply]
Ok this could work but in that case the `aqua.yml` has to be a link to a common aqua.yaml (like that if the file has to be updated for all team it is done once). That's why having some kind of inheritance would be easier:
```
/etc/aqua.d/
  aqua.yaml
  aqua-checksums.json
  teamA.d/
    aqua.yaml
    aqua-checksums.json
  teamB.d/
    aqua.yaml
    aqua-checksums.json
```
aqua.yaml
```
checksum:
  enabled: true
  require_checksum: true
  supported_envs:
    - linux/amd64
registries:
  - type: standard
    ref: v4.138.0 # renovate: depName=aquaproj/aqua-registry
packages:
  - import: "*.d/*.yml"
 ```
 teamA/aqua.yml
```
---
packages:
  - name: helm/helm@v3.14.0
    description: The Kubernetes Package Manager
    link: https://github.com/helm/helm
```
teamA/aqua-checksums.json
```
{
  "checksums": [
    {
      "id": "http/get.helm.sh/helm-v3.14.0-linux-amd64.tar.gz",
      "checksum": "F43E1C3387DE24547506AB05D24E5309C0CE0B228C23BD8AA64E9EC4B8206651",
      "algorithm": "sha256"
    }
  ]
}
```
And just doing a deep merge would help.
But I can understand it would bring complexity.

On a larger point of view, having inheritance for every top level (registry, packages, checksums) could easy this kind of setup

**Reply by suzuki-shunsuke** [reply]
> the aqua.yml has to be a link to a common aqua.yaml (like that if the file has to be updated for all team it is done once).

Sorry, what do you mean?
As I mentioned you can specify multiple global configuration files, so you just update common aqua.yaml then the update would be reflected to all teams. https://github.com/orgs/aquaproj/discussions/2699#discussioncomment-8550478

Generally speaking, the inheritance has both pros and cons.
The cons is that it is difficult to understand the merged settings and the implementation gets complicated.
To understand the merged settings, we need to check all files (I talk about not only aqua but also more general stuffs).
So we avoid the inheritance unless the inheritance is really necessary when we develop tools.

**Reply by jmauro** [reply]
Does not seem to work the way I understood it:
```
$ export AQUA_GLOBAL_CONFIG=/etc/aqua/team.d/aqua.yml:/etc/aqua/aqua.yml
$ cat /etc/aqua/team.d/aqua.yml
---
packages:
  - name: aws/aws-cli@2.15.21
    description: The AWS Command Line Interface (AWS CLI) is a unified tool that provides a consistent interface for interacting with all parts of Amazon Web Services
    link: https://github.com/aws/aws-cli
$ cat /etc/aqua/aqua.yml
---
# aqua - Declarative CLI Version Manager
# https://aquaproj.github.io/
checksum:
  enabled: true
  require_checksum: true
  supported_envs:
    - linux/amd64
registries:
  - type: standard
    ref: v4.137.0 # renovate: depName=aquaproj/aqua-registry
  - name: test
    type: local
    path: /opt/registry/registry.yaml
packages:
  - import: "*.d/*.yml"
  - import: "*.d/*.yaml"
 $ cd /etc/aqua/team.d/
$ aqua install -a
aqua install -a
ERRO[0000] install the package                           aqua_version=2.23.1 env=linux/amd64 error="registry isn't found" package_name=aws/aws-cli package_version=2.15.21 program=aqua registry=standard
FATA[0000] aqua failed                                   aqua_version=2.23.1 env=linux/amd64 error="it failed to install some packages" program=aqua
```
Like I said, I thought there was some kind of inheritance, but it is not the case.

Thanks for taking the time to help me.

**Reply by suzuki-shunsuke** [reply]
```yaml
$ cat /etc/aqua/team.d/aqua.yml
---
packages:
  - name: aws/aws-cli@2.15.21
    description: The AWS Command Line Interface (AWS CLI) is a unified tool that provides a consistent interface for interacting with all parts of Amazon Web Services
    link: https://github.com/aws/aws-cli
```

```
ERRO[0000] install the package                           aqua_version=2.23.1 env=linux/amd64 error="registry isn't found" package_name=aws/aws-cli package_version=2.15.21 program=aqua registry=standard
```

You specified `/etc/aqua/team.d/aqua.yml` in `AQUA_GLOBAL_CONFIG`, so `registries` is required.
We can omit `registries` if the file is imported.

**Comment by suzuki-shunsuke** [comment]
BTW, I have a trivial suggestion.
In my understanding, you manage tools with aqua on a specific server, right?
If so, you can specify not only os but also arch in `checksum.supported_envs`.

```yaml
checksum:
  enabled: true
  require_checksum: true
  supported_envs:
    - linux/amd64 # or linux/arm64
```

Then aqua gets checksums of only the given os and arch.
This makes aqua-checksums.json slim and prevents aqua from getting unnecessary checksums.

**Reply by jmauro** [reply]
Thanks for the tips (it helps)
